dandi/dandiset.py

```diff
@@ -131,20 +145,51 @@ def identifier(self) -> str:
             )
         return id_
 
-
-class APIDandiset(Dandiset):
-    """A dandiset to replace "classical" Dandiset whenever we migrate to new API based server"""
-
-    def __init__(
-        self,
-        path: Union[str, Path],
-        allow_empty: bool = False,
-        schema_version: Optional[str] = None,
-    ) -> None:
-        if schema_version is not None:
-            current_version = get_schema_version()
-            if schema_version != current_version:
-                raise ValueError(
-                    f"Unsupported schema version: {schema_version}; expected {current_version}"
-                )
-        super().__init__(path, allow_empty=allow_empty)
+    def assets(self, allow_all: bool = False) -> AssetView:
+        data = {}
+        for df in find_dandi_files(
+            self.path, dandiset_path=self.path, allow_all=allow_all
+        ):
+            if isinstance(df, DandisetMetadataFile):
+                continue
+            assert isinstance(df, LocalAsset)
+            data[PurePosixPath(df.path)] = df
+        return AssetView(data)
+
+    def metadata_file(self) -> DandisetMetadataFile:
+        df = dandi_file(self._metadata_file_obj, dandiset_path=self.path)
+        assert isinstance(df, DandisetMetadataFile)
+        return df
+
+
+@dataclass
+class AssetView:
+    """
+    A collection of all assets in a local Dandiset, used to ensure that
+    `BIDSDatasetDescriptionAsset` objects are stored and remain alive while
+    working with only a subset of the files in a Dandiset.
+    """
+
+    data: dict[PurePosixPath, LocalAsset]
+
+    def __iter__(self) -> Iterator[LocalAsset]:
+        return iter(self.data.values())
+
+    def under_paths(self, paths: Iterable[str | Path]) -> Iterator[LocalAsset]:
+        # The given paths must be relative to the Dandiset root.
+        ppaths = [PurePosixPath(p) for p in paths]
+        for p, df in self.data.items():
+            if any(is_relative_to(p, pp) for pp in ppaths):
+                yield df
```

Comment:
this would make it N*2 operation  which could become prohibitively slow in large dandisets and use of globs (e.g. (e.g. imagine a dataset with 1,000 subjects folders and doing matching for thousand paths under them --  order of 1,000,000 iterations)... but hey -- even in DataLad we have code paths were we do that and still have OPT comment, e.g.
https://github.com/datalad/datalad/blob/HEAD/datalad/interface/utils.py#L121 although I believe we even have some helpers already to make it `O(N*log(N))` e.g. [`get_parent_paths`](https://github.com/datalad/datalad/blob/HEAD/datalad/support/path.py#L125) and more directly applicable here [`get_filtered_paths_`](https://github.com/datalad/datalad/blob/HEAD/datalad/support/path.py#L216).

@jwodder -- do you think we could adopt that `get_filtered_paths_` within our `utils` and use here?  or may be you could offer some cleaner solution? 

surprisingly quick googling did not lead me to some efficient answer, including this [SO post](https://stackoverflow.com/questions/68991483/using-pathlib-path-how-to-efficiently-check-if-a-directory-resides-anywhere-und) doesn't have it either

Reply:
Done: 510fce7.